OpenSim/Simulation/Model/OffsetFrame.cpp

```diff
@@ -0,0 +1,209 @@
+    setOffsetTransform(offset);
+}
+//_____________________________________________________________________________
+/**
+* Set a null frame as Identity rotation, 0 translation
+*/
+template <class C>
+void OffsetFrame<C>::setNull()
+{
+    _offsetTransform.setToNaN();
+    this->setAuthors("Matt DeMers, Ajay Seth");
+}
+//_____________________________________________________________________________
+/**
+* construct properties
+*/
+template <class C>
+void OffsetFrame<C>::constructProperties()
+{
+    SimTK::Vec3 zero(0.0, 0.0, 0.0);
+    constructProperty_translation(zero);
+    constructProperty_orientation(zero);
+    // transform at default
+}
+
+template <class C>
+void OffsetFrame<C>::constructConnectors()
+{
+    this->template constructConnector<C>("parent");
+}
+
+//=============================================================================
+// FRAME COMPUTATIONS
+//=============================================================================
+//_____________________________________________________________________________
+/*
+ * Implementation of Frame interface by OffsetFrame.
+ *
+ */
+template <class C>
+const SimTK::Transform& OffsetFrame<C>::
+    calcGroundTransform(const SimTK::State& s) const
+{
+    if (!this->isCacheVariableValid(s, "ground_transform")){
+        this->template setCacheVariableValue<SimTK::Transform>(s,
+                "ground_transform",
+                getParentFrame().getGroundTransform(s)*getOffsetTransform());
+    }
+    // return X_GF = X_GB * X_BF where F is the offset frame;
+    return this->template getCacheVariableValue<SimTK::Transform>(s,
+            "ground_transform");
+}
+
+//=============================================================================
+// GET AND SET
+//=============================================================================
+//_____________________________________________________________________________
+template <class C>
+void OffsetFrame<C>::setParentFrame(const C& parent)
+{ 
+    this->template updConnector<C>("parent").connect(parent);
+}
+
+template <class C>
+const C& OffsetFrame<C>::getParentFrame() const
+{
+    return this->template getConnector<C>("parent").getConnectee();
+}
+
+template <class C>
+const SimTK::Transform& OffsetFrame<C>::getOffsetTransform() const
+{
+    return _offsetTransform;
+}
+
+template <class C>
+void OffsetFrame<C>::setOffsetTransform(const SimTK::Transform& xform)
+{
+    _offsetTransform = xform;
+    // Make sure properties are updated in case we either call gettters or
+    // serialize after this call
+    set_translation(xform.p());
+    set_orientation(xform.R().convertRotationToBodyFixedXYZ());
+}
+
+template<class C>
+const Frame& OffsetFrame<C>::extendFindBaseFrame() const
+{
+    // check if parent is also an offset 
+    const OffsetFrame<C>* parentOffset =
+        dynamic_cast<const OffsetFrame<C>*>(&getParentFrame());
+    if (parentOffset) {
+        return parentOffset->findBaseFrame();
```

Comment:
This doesn't look good to me. It is kind of faking up a virtual method call in a very expensive and awkward way (dynamic_cast is much slower than a virtual function call). This suggests to me that the base class should probably have a virtual `findBaseFrame()` method.